### PR TITLE
Capi: Composite Method enhancement

### PR DESCRIPTION
**Pull request by patrykka**
- Description :

- Issue Tickets (if any) :

- Tests or Samples (if any) :

- Screen Shots (if any) :

### CODE REVIEW

File: inc/thorvg_capi.h [hunk 1]
```diff
@@ -579,15 +579,27 @@ TVG_EXPORT Tvg_Paint* tvg_paint_duplicate(Tvg_Paint* paint);
...
+/*!
+* \fn TVG_EXPORT Tvg_Result tvg_paint_get_composite_method(Tvg_Paint* paint, Tvg_Composite_Method **method)
+* \brief The function get composition method
+* \param[in] paint Tvg_Paint pointer
+* \param[in] method Tvg_Composite_Method used composite method
+* \return Tvg_Result return value
+* - TVG_RESULT_SUCCESS: if ok.
+* - TVG_RESULT_INVALID_PARAMETERS: if paint is invalid
+*/
+TVG_EXPORT Tvg_Result tvg_paint_get_composite_method(Tvg_Paint* paint, Tvg_Composite_Method **method);
```
Comment (hermet) [inline inc/thorvg_capi.h:602]: => const Tvg_Paint* paint,

File: inc/thorvg_capi.h [hunk 2]
```diff
@@ -579,15 +579,27 @@ TVG_EXPORT Tvg_Paint* tvg_paint_duplicate(Tvg_Paint* paint);
...
 * \brief The function set composition method
 * \param[in] paint Tvg_Paint pointer
+* \param[in] mask Tvg_Paint pointer to mask object
 * \param[in] method Tvg_Composite_Method used composite method
 * \return Tvg_Result return value
 * - TVG_RESULT_SUCCESS: if ok.
 * - TVG_RESULT_INVALID_PARAMETERS: if paint is invalid
 */
-TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(const Tvg_Paint* paint, Tvg_Composite_Method method);
+TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(const Tvg_Paint* paint, const Tvg_Paint* mask, Tvg_Composite_Method method);
```
Comment (hermet) [inline inc/thorvg_capi.h:591]: here const in parameters shouldn't be here. const is declared only when the argument is not changed.

**Review by hermet — CHANGES_REQUESTED** [review]
please check comments.

File: inc/thorvg_capi.h [hunk 3]
```diff
@@ -579,15 +579,27 @@ TVG_EXPORT Tvg_Paint* tvg_paint_duplicate(Tvg_Paint* paint);
 TVG_EXPORT Tvg_Result tvg_paint_get_bounds(const Tvg_Paint* paint, float* x, float* y, float* w, float* h);
 
 /*!
-* \fn TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(const Tvg_Paint* paint, Tvg_Composite_Method method)
+* \fn TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(const Tvg_Paint* paint, const Tvg_Paint* mask, Tvg_Composite_Method method)
```
Comment (hermet) [inline inc/thorvg_capi.h:582]: --const

**Review by hermet — CHANGES_REQUESTED** [review]
please check a comment, 
+ question, Any necessity of getter interface?

File: inc/thorvg_capi.h (hunk 3 above)
Comment (patrykka) [reply inc/thorvg_capi.h]: After you ask your question about getter, I realized that in fact I dont need it.
Removed.

File: inc/thorvg_capi.h [hunk 4]
```diff
@@ -579,15 +579,17 @@ TVG_EXPORT Tvg_Paint* tvg_paint_duplicate(Tvg_Paint* paint);
 TVG_EXPORT Tvg_Result tvg_paint_get_bounds(const Tvg_Paint* paint, float* x, float* y, float* w, float* h);
```
Comment (mihashco) [inline inc/thorvg_capi.h:580]: Please keep API new lines consistency.

File: inc/thorvg_capi.h [hunk 5]
```diff
@@ -579,15 +579,17 @@ TVG_EXPORT Tvg_Paint* tvg_paint_duplicate(Tvg_Paint* paint);
...
 * \param[in] paint Tvg_Paint pointer
+* \param[in] mask Tvg_Paint pointer to mask object
 * \param[in] method Tvg_Composite_Method used composite method
 * \return Tvg_Result return value
 * - TVG_RESULT_SUCCESS: if ok.
 * - TVG_RESULT_INVALID_PARAMETERS: if paint is invalid
 */
-TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(const Tvg_Paint* paint, Tvg_Composite_Method method);
+TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(Tvg_Paint* paint, Tvg_Paint* mask, Tvg_Composite_Method method);
+
```
Comment (mihashco) [inline inc/thorvg_capi.h:592]: Only one new line is used before new section

**Review by mihashco — CHANGES_REQUESTED** [review]

**Comment by hermet** [comment]
@patrykka question, Any necessity of getter interface?

**Comment by patrykka** [comment]
> @patrykka question, Any necessity of getter interface?

@hermet
Use case is on tizen.org. I created patch-set in efl project and in elm-demo example, both on tizen branches.
You were thinkng about different method of composite method setter ?

**Review by mihashco — APPROVED** [review]

**Comment by patrykka** [comment]
@hermet could you please verify this again?
Currently, setter interface is adding composition method to the same object which is composition target:

``` c++
return (Tvg_Result) reinterpret_cast<const Paint*>(paint)->composite(unique_ptr<Paint>((Paint*)(paint)), (CompositeMethod)method);
```
change is fixing this issue:
``` c++
return (Tvg_Result) reinterpret_cast<Paint*>(paint)->composite(unique_ptr<Paint>((Paint*)(comp_obj)), (CompositeMethod)method);
```

File: inc/thorvg_capi.h [hunk 6]
```diff
@@ -578,16 +578,18 @@ TVG_EXPORT Tvg_Paint* tvg_paint_duplicate(Tvg_Paint* paint);
 */
 TVG_EXPORT Tvg_Result tvg_paint_get_bounds(const Tvg_Paint* paint, float* x, float* y, float* w, float* h);
 
+
 /*!
-* \fn TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(const Tvg_Paint* paint, Tvg_Composite_Method method)
+* \fn TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(Tvg_Paint* paint, Tvg_Paint* mask, Tvg_Composite_Method method)
```
Comment (hermet) [inline inc/thorvg_capi.h:583]: mask => target

File: inc/thorvg_capi.h [hunk 7]
```diff
@@ -578,16 +578,18 @@ TVG_EXPORT Tvg_Paint* tvg_paint_duplicate(Tvg_Paint* paint);
...
 * \brief The function set composition method
 * \param[in] paint Tvg_Paint pointer
+* \param[in] comp_obj Tvg_Paint pointer to composition object
 * \param[in] method Tvg_Composite_Method used composite method
 * \return Tvg_Result return value
 * - TVG_RESULT_SUCCESS: if ok.
 * - TVG_RESULT_INVALID_PARAMETERS: if paint is invalid
 */
-TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(const Tvg_Paint* paint, Tvg_Composite_Method method);
+TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(Tvg_Paint* paint, Tvg_Paint* comp_obj, Tvg_Composite_Method method);
```
Comment (hermet) [inline inc/thorvg_capi.h:592]: comp_obj => target

File: src/bindings/capi/tvgCapi.cpp [hunk 8]
```diff
@@ -191,10 +191,11 @@ TVG_EXPORT Tvg_Result tvg_paint_get_bounds(const Tvg_Paint* paint, float* x, flo
    return (Tvg_Result) reinterpret_cast<const Paint*>(paint)->bounds(x, y, w, h);
 }
 
-TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(const Tvg_Paint* paint, Tvg_Composite_Method method)
+
+TVG_EXPORT Tvg_Result tvg_paint_set_composite_method(Tvg_Paint* paint, Tvg_Paint* comp_obj, Tvg_Composite_Method method)
 {
    if (!paint) return TVG_RESULT_INVALID_ARGUMENT;
-   return (Tvg_Result) reinterpret_cast<const Paint*>(paint)->composite(unique_ptr<Paint>((Paint*)(paint)), (CompositeMethod)method);
+   return (Tvg_Result) reinterpret_cast<Paint*>(paint)->composite(unique_ptr<Paint>((Paint*)(comp_obj)), (CompositeMethod)method);
```
Comment (hermet) [inline src/bindings/capi/tvgCapi.cpp:198]: comp_obj => target

**Review by hermet — CHANGES_REQUESTED** [review]
Here is the prototype, we don't need to change the name.

Result composite(std::unique_ptr<Paint> target, CompositeMethod method) const noexcept;

**Pull request closed without merging**